modules/apps/27-interchain-accounts/host/keeper/msg_server.go

```diff
@@ -28,37 +30,63 @@ func NewMsgServerImpl(keeper *Keeper) types.MsgServer {
 
 // ModuleQuerySafe routes the queries to the keeper's query router if they are module_query_safe.
 // This handler doesn't use the signer.
-func (m msgServer) ModuleQuerySafe(goCtx context.Context, msg *types.MsgModuleQuerySafe) (*types.MsgModuleQuerySafeResponse, error) {
-	ctx := sdk.UnwrapSDKContext(goCtx)
-
+func (m msgServer) ModuleQuerySafe(ctx context.Context, msg *types.MsgModuleQuerySafe) (*types.MsgModuleQuerySafeResponse, error) {
 	responses := make([][]byte, len(msg.Requests))
 	for i, query := range msg.Requests {
 		isModuleQuerySafe := slices.Contains(m.mqsAllowList, query.Path)
 		if !isModuleQuerySafe {
 			return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidRequest, "not module query safe: %s", query.Path)
 		}
 
-		route := m.queryRouter.Route(query.Path)
-		if route == nil {
-			return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidRequest, "no route to query: %s", query.Path)
+		path := strings.TrimPrefix(query.Path, "/")
+		pathFullName := protoreflect.FullName(strings.ReplaceAll(path, "/", "."))
+
+		desc, err := gogoproto.GogoResolver.FindDescriptorByName(pathFullName)
+		if err != nil {
+			return nil, err
+		}
+
+		md, isGRPC := desc.(protoreflect.MethodDescriptor)
+		if !isGRPC {
+			return nil, errorsmod.Wrapf(ibcerrors.ErrInvalidRequest, "no descriptor found for query path: %s", string(desc.FullName()))
 		}
 
-		res, err := route(ctx, &abci.QueryRequest{
-			Path: query.Path,
-			Data: query.Data,
-		})
+		msg, err := forgeProtoTypeFromName(string(md.Input().FullName()))
 		if err != nil {
-			m.Logger(ctx).Debug("query failed", "path", query.Path, "error", err)
 			return nil, err
 		}
-		if res == nil || res.Value == nil {
+
+		if err := m.cdc.Unmarshal(query.Data, msg); err != nil {
+			return nil, err
```

Comment:
I think it might be best to return a more "constant" error here if this rpc is intended to be run in OnRecvPacket. Protobuf unmarshalling errors have a possibility to be indeterminate (this was a something found in earlier days of ibc transfer).

cc. @AdityaSripal this is correct, right?

Reply:
Actually, I think this is protected by the fact that we changed acknowledgement errors themselves to strip out abci error code alone. But none the less, this is something good to be aware of cc. @gjermundgaraba

Reply:
for example, back in the day we used to do this https://github.com/cosmos/ibc-go/blob/v3.4.0/modules/apps/transfer/ibc_module.go#L177 and it was for that reason.